kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+// kzip archive, and returns lists of their relative paths and FileInput protos.
+func collectCrateSourcesImpl(ctx context.Context, e *extractor, sourceDirs source) ([]string, []*apb.CompilationUnit_FileInput, error) {
+	var sourceFiles []string
+	var requiredInputs []*apb.CompilationUnit_FileInput
+
+	// cache required inputs for each include_dir
+	for _, includeDir := range sourceDirs.IncludeDirs {
+		if e.requiredInputsCache.inputsCache[includeDir] != nil {
+			requiredInputs = append(requiredInputs, e.requiredInputsCache.inputsCache[includeDir]...)
+			sourceFiles = append(sourceFiles, e.requiredInputsCache.sourceFilesCache[includeDir]...)
+
+			// we don't need to add this file to the kzip_writer, because we know it's already been written.
+			continue
+		}
+
+		dirRequiredInputs := []*apb.CompilationUnit_FileInput{}
+		err := vfs.Walk(ctx, filepath.Join(e.projectRoot, includeDir), func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			shouldInclude, err := shouldIncludeFile(info, path, sourceDirs, e.projectRoot)
+			if err != nil || !shouldInclude {
+				return nil
+			}
+
+			relativePath, err := removeProjectRoot(path, e.projectRoot)
+			if err != nil {
+				return err
+			}
+
+			input, err := vfs.Open(ctx, path)
+			if err != nil {
+				return err
+			}
+			defer input.Close()
+
+			digest, err := e.kzipWriter.AddFile(input)
+			if err != nil {
+				return err
+			}
+
+			vname := &spb.VName{
+				Corpus: e.corpus,
+				Path:   relativePath,
+			}
+
+			if inferredVname, ok := e.rules.Apply(relativePath); ok {
+				vname = inferredVname
+			}
+
+			sourceFiles = append(sourceFiles, relativePath)
+			dirRequiredInputs = append(dirRequiredInputs, &apb.CompilationUnit_FileInput{
+				VName: vname,
+				Info: &apb.FileInfo{
+					Path:   relativePath,
+					Digest: digest,
+				},
+			})
+			return nil
+		})
+
+		if err != nil {
+			continue
+		}
+
+		requiredInputs = append(requiredInputs, dirRequiredInputs...)
+		e.requiredInputsCache.inputsCache[includeDir] = dirRequiredInputs
+		e.requiredInputsCache.sourceFilesCache[includeDir] = sourceFiles
+	}
+
+	return sourceFiles, requiredInputs, nil
+}
+
+// synthesizeProjectJson creates a RustProject from a given crate and its transitive dependencies
+func synthesizeProjectJson(crateId crateId, crates []crate, transitiveDeps map[crateId][]crateId) rustProject {
+	project := rustProject{
+		Crates: []crate{},
+	}
+	currentCrate := crates[crateId]
+
+	// override IsWorkspaceMember
+	currentCrate.IsWorkspaceMember = true
+
+	project.Crates = append(project.Crates, currentCrate)
+
+	for _, dep := range transitiveDeps[crateId] {
+		// override IsWorkspaceMember
+		crate := crates[dep]
+		crate.IsWorkspaceMember = true
+		project.Crates = append(project.Crates, crate)
+	}
+	sort.Slice(project.Crates, func(i, j int) bool {
+		return project.Crates[i].CrateId < project.Crates[j].CrateId
```

Comment:
nit: I'd suggest sorting this list rather where the list of transitive deps is defined, that's less distance to plumb the required index over.
(e.g. if you implement sort.Interface then you can use external indexes and don't need to include CrateId at all)

alternatively, the BFS traversal order *is* deterministic so could be good enough. (sorting is a bit nicer though)

Reply:
I think I've addressed this by keeping crate ID, which worked.